core/types/request.go

```diff
@@ -0,0 +1,162 @@
+// Copyright 2024 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package types
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"io"
+
+	"github.com/ethereum/go-ethereum/rlp"
+)
+
+var (
+	ErrRequestTypeNotSupported = errors.New("request type not supported")
+	errShortTypedRequest       = errors.New("typed request too short")
+)
+
+// Request types.
+const (
+	DepositRequestType = 0x00
+)
+
+// Request is an EIP-7685 request object. It represents execution layer
+// triggered messages bound for the consesus layer.
+type Request struct {
+	inner RequestData
+}
+
+// Type returns the EIP-7685 type of the request.
+func (r *Request) Type() byte {
+	return r.inner.requestType()
+}
+
+// NewRequest creates a new request.
+func NewRequest(inner RequestData) *Request {
+	req := new(Request)
+	req.inner = inner.copy()
+	return req
+}
+
+// Requests implements DerivableList for requests.
+type Requests []*Request
+
+// Len returns the length of s.
+func (s Requests) Len() int { return len(s) }
+
+// EncodeIndex encodes the i'th request to s.
+func (s Requests) EncodeIndex(i int, w *bytes.Buffer) {
+	s[i].encode(w)
+}
+
+// Retrieve deposits from a requests list.
+func (s Requests) Deposits() Deposits {
+	deposits := make(Deposits, 0, len(s))
+	for _, req := range s {
+		if req.Type() == DepositRequestType {
+			deposits = append(deposits, req.inner.(*Deposit))
+		}
+	}
+	return deposits
+}
+
+type RequestData interface {
+	requestType() byte
+	encode(*bytes.Buffer) error
+	decode([]byte) error
+	copy() RequestData // creates a deep copy and initializes all fields
+}
+
+// EncodeRLP implements rlp.Encoder
+func (r *Request) EncodeRLP(w io.Writer) error {
+	buf := encodeBufferPool.Get().(*bytes.Buffer)
+	defer encodeBufferPool.Put(buf)
+	buf.Reset()
+	if err := r.encode(buf); err != nil {
+		return err
+	}
+	return rlp.Encode(w, buf.Bytes())
+}
+
+// encode writes the canonical encoding of a request to w.
+func (r *Request) encode(w *bytes.Buffer) error {
+	w.WriteByte(r.Type())
+	return r.inner.encode(w)
+}
+
+// MarshalBinary returns the canonical encoding of the request.
+func (r *Request) MarshalBinary() ([]byte, error) {
```

Comment:
This panics if the inner is not set, I guess it is okay since we are only ever marshaling well formed requests

Reply:
Exactly. This is similar to transactions also.